src/hangar/cli/cli.py

```diff
@@ -359,136 +360,172 @@ def server(overwrite, ip, port, timeout):
               help='overwrite data samples with the same name as the imported data file ')
 @pass_repo
-def import_data(repo: Repository, arrayset, path, branch, plugin, overwrite):
+@click.pass_context
+def import_data(ctx, repo: Repository, arrayset, path, branch, plugin, overwrite):
     """Import file(s) at PATH to ARRAYSET in the staging area.
     """
-    from hangar.cli.io import imread
+    # TODO: ignore warning through env variable
+    import types
+    from hangar import external
     from hangar.records.heads import get_staging_branch_head
+    kwargs = parse_custom_arguments(ctx.args)
 
-    try:
-        if branch is not None:
-            if branch in repo.list_branches():
-                branch_name = branch
-            else:
-                click.echo(f'Branch name: {branch} does not exist, Exiting.')
-                return None
+    if branch is not None:
+        if branch in repo.list_branches():
+            branch_name = branch
         else:
-            branch_name = get_staging_branch_head(repo._env.branchenv)
-        click.echo(f'Writing to branch: {branch_name}')
-
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore", UserWarning)
-            co = repo.checkout(write=True, branch=branch_name)
-            aset = co.arraysets.get(arrayset)
-
-        if os.path.isfile(path):
-            fname = os.path.basename(path)
-            if not overwrite:
-                if fname in aset:
-                    click.echo(f'skipping existing name: {fname} as overwrite flag not set')
-                    return None
-            fNamePth = [(fname, path)]
+            click.echo(f'Branch name: {branch} does not exist, Exiting.')
+            return None
+    else:
+        branch_name = get_staging_branch_head(repo._env.branchenv)
+    click.echo(f'Writing to branch: {branch_name}')
+
+    co = repo.checkout(write=True, branch=branch_name)
+    try:
+        aset = co.arraysets.get(arrayset)
+        if os.path.isdir(path):
+            files = [os.path.join(path, fname) for fname in os.listdir(path)]
         else:
-            fnames = os.listdir(path)
-            if not overwrite:
-                fnames = [fname for fname in fnames if fname not in aset]
-            fNamePth = [(fname, os.path.join(path, fname)) for fname in fnames]
-
-        with aset as a, click.progressbar(fNamePth) as fnamesBar:
-            for fn, fpth in fnamesBar:
-                arr = imread(fpth, plugin=plugin)
-                try:
-                    a[fn] = arr
-                except ValueError as e:
-                    click.echo(e)
+            files = [path]
+        with aset, click.progressbar(files) as filesBar:
+            for fpath in filesBar:
+                extension = os.path.splitext(fpath)[1].strip('.')
+                loaded = external.load(fpath, plugin=plugin, extension=extension, **kwargs)
+                if isinstance(loaded, types.GeneratorType):
```

Comment:
If we are always passing in a single file path (and not a directory), why would this be a `GeneratorType`? wouldn't  "one file in" always result in "one piece of data out" ?

Reply:
If you meant how are we handling the single file case, that's the `else`

Reply:
Ahh ok, i was a bit confused here. It wasn't immediately clear that this was designed to handle situations where a file can output many different samples to store in an arrayset. Highlighting this with something like a CSV file (in a test or comment) would be helpful for future readers..

Reply:
Good point, Will do that